Add support for handling multiple identify results

nCastle1 opened the pull request:
Changes:

* Now supports multiple identify results
* Now supports identifying polygons
* An arcade-based subtitle expression is used to differentiate multiple identified features
* A subtle bug in the custom layout panel has been fixed

nCastle1 committed:
Adds support for multiple identify results

nCastle1 committed:
Fixes subtle layout issue

nCastle1 committed:
Remove unused references

esreli commented on src/DataCollection.Shared/ViewModels/FeatureViewModel.cs:
```diff
@@ -303,6 +309,45 @@ internal void RaiseFeatureCRUDOperationCompleted(CRUDOperation operation)
+            }
+        }
+
+        private string _subtitle = "...";
```
Is this the default arcade expression? Should we make the default subtitle string empty? What's the fallback plan here?

```suggestion
        private string _subtitle = "";
```

esreli commented on src/DataCollection.Shared/ViewModels/IdentifyResultViewModel.cs:
```diff
@@ -0,0 +1,175 @@
+            }
+
+            // Select the only feature if there is only one feature
+            if (IdentifiedFeatures != null && IdentifiedFeatures.Count() == 1)
```
Does this case skip showing the results list view in favor of showing the single result detail view?

nCastle1 replied:
Yes, if there is only one result, that one result is shown directly. I figured that would save time and clicks in cases where there is no ambiguity, but that can be changed if we want to show the navigation every time.

esreli replied:
I think the behavior described is good 👍

esreli commented on src/DataCollection.Shared/ViewModels/IdentifyResultViewModel.cs:
```diff
@@ -0,0 +1,175 @@
+        }
+
+        /// <summary>
+        /// Gets the command that selects the next result in the list, with wraparound.
```
nice

esreli commented on RELEASE.md:
```diff
@@ -8,6 +8,7 @@
     * Switches to vector icons where possible.
 * Moves the .NET Core WPF project to a separate folder to prevent possible build issues.
 * Updates the code called at sign out to revoke credentials from the server, when possible, and delete them locally.
+* Adds support for identifying mulitple items on the map simultaneously. A unique interface gives you enough context to choose a result, with context display configurable via the web map.
```
I'm not really sure what this means, can you explain it? Perhaps we remove it altogether?

> A unique interface gives you enough context to choose a result, with context display configurable via the web map.

Maybe you meant something like,

"Users can interrogate the list to display a feature's full details."

```suggestion
* Adds support for identifying multiple features on the map and displaying them in a list.
```

nCastle1 replied:
I was referencing the subtitle feature indirectly, maybe it makes sense to call that out more explicitly? It is a somewhat unique pattern that I haven't seen elsewhere.

esreli commented on docs/README.md:
```diff
@@ -380,6 +386,16 @@ These rules are wrapped conveniently as extensions into a static class named `Ap
 
+### Multiple identify results
+
+When the user clicks or taps on the map, all of the map's layers and all of the map view's graphics overlays are queried. The results, if there are more than one, are shown in a list. The list shows each result's graphical representation (if applicable), popup title, and subtitle.
```
```suggestion
When the user clicks or taps on the map, all of the map's layers and all of the map view's graphics overlays are queried. The results are displayed in a list showing each result's graphical representation, popup title, and configured subtitle.
```

esreli commented on docs/README.md:
```diff
@@ -380,6 +386,16 @@ These rules are wrapped conveniently as extensions into a static class named `Ap
+
+When the user clicks or taps on the map, all of the map's layers and all of the map view's graphics overlays are queried. The results, if there are more than one, are shown in a list. The list shows each result's graphical representation (if applicable), popup title, and subtitle.
+
+> **Note**: Subtitles are not a standard part of the ArcGIS platform and require special attention to set up.
```
```suggestion
> **Note**, subtitles are not a standard part of the ArcGIS platform and require special attention to set up.
```

esreli commented on docs/README.md:
```diff
@@ -380,6 +386,16 @@ These rules are wrapped conveniently as extensions into a static class named `Ap
+
+> **Note**: Subtitles are not a standard part of the ArcGIS platform and require special attention to set up.
+
+The subtitle is computed via an Arcade expression defined in the pop-up configuration's *Attribute Expressions* collection. To be displayable in the app, the attribute expression must have the alias or title set to `Subtitle expression`.
```
Can you link to ArcGIS documentation for Arcade expressions?

```suggestion
The subtitle is computed via an Arcade expression defined in the pop-up configuration's *Attribute Expressions* collection. The app will only display a subtitle if the attribute expression has the alias or title set to `Subtitle expression`.
```

esreli commented on docs/README.md:
```diff
@@ -380,6 +386,16 @@ These rules are wrapped conveniently as extensions into a static class named `Ap
+
+The subtitle is computed via an Arcade expression defined in the pop-up configuration's *Attribute Expressions* collection. To be displayable in the app, the attribute expression must have the alias or title set to `Subtitle expression`.
+
+> **Note**: The `Subtitle expression` is found by an exact string match. Do not change the title when localizing the web map or for any other reason.
```
```suggestion
> **Note**, the `Subtitle expression` is found by an exact string match. Do not change the title when localizing the web map or for any other reason.
```

esreli commented on src/DataCollection.Shared/Binding Support/IdentifyController.cs:
```diff
@@ -184,7 +184,7 @@ private async void MapView_Tapped(object sender, GeoViewInputEventArgs e)
                 var pixelTolerance = 10;
                 var returnPopupsOnly = false;
-                var maxResultCount = 5;
+                var maxResultCount = 8;
```
Is this per layer? Is that determined to be an appropriate amount? 

Further, perhaps we consider placing this and maybe other constants in a configuration file?

esreli commented on src/DataCollection.Shared/Converters/NumberToResultCountConverter.cs:
```diff
@@ -0,0 +1,45 @@
+        public object Convert(object value, Type targetType, object parameter, CustomCultureInfo culture)
+        {
+
+            return $"{value} {Resources.GetString(parameter.ToString())}";
```
```suggestion
            return $"{value} {Resources.GetString(parameter.ToString())}";
```

esreli commented on src/DataCollection.Shared/ViewModels/FeatureViewModel.cs:
```diff
@@ -94,7 +96,9 @@ private async void UpdateImageSource()
                 }
 
-                var symbol = await PopupManager.Symbol.CreateSwatchAsync(16, 16, 192, System.Drawing.Color.Transparent);
+                // Use different sizes depending on the geometry type to get best results.
```
We've spoken about this because it feels a bit hacky and too customized to our dataset... are we sure this is the best method?  Should we not allow the SDK to derive the symbol based on its inherent size? I'm afraid of some downstream unintended consequences for symbols from other data authors whose sizes are meticulously calculated.

`public virtual Task<RuntimeImage> CreateSwatchAsync( double dpi = 96 )`

esreli reviewed (changes requested):
I love seeing this implementation used in the new layout system. Very cool!

### Results counter

The font size used to convey the results counter ("8 of 8") is the same size as the pop-up title. Can we decrease to a font size closer to that of a subtitle or footer?

<img width="402" alt="Screen Shot 2020-10-02 at 10 27 54 AM" src="https://user-images.githubusercontent.com/33433113/94951907-f1e8b800-0499-11eb-834e-3b96f62f3b95.png">

### Map identify rules

Have you evaluated if the application still needs this rule at all? I'm wondering if we can remove this code (& subsequent docs section.)

### Graphics Overlays queried

It appears this PR introduces support for identifying graphic overlays, I have a few questions here. Namely how do we support graphic overlays? More specifically,

1. How would this application receive graphics to begin with? They aren't provided by the web map service. In what scenario would identifying a graphics overlay yield any results?
1. As I understand it, graphics overlays will not be (cannot be?) configured with a pop-up configuration, how will the pop-up list cell & detail view know what content to display and in what order?
1. The app supports editing pop-ups, how is editing supported with graphic overlays?

My intuition here is that without full support for graphic overlays, we should exclude them from the map identify operation.

### Subtitle expression

I like the idea you've come up using an arcade expression defined in the web map to drive the subtitle. A few thoughts here.

1. Can we consider a different static string other than `'Subtitle expression'`? If this expression is tied to the layer, perhaps we simply go with `'subtitle'`? This might be an easier standard for our customers to adopt.
1. I'm wondering if we should consider sticking this static string in a Configuration file? This way it is easier to change alongside any other web map configs.

nCastle1 committed:
Apply suggestions from code review

Co-authored-by: Eli Gregory <33433113+esreli@users.noreply.github.com>

nCastle1 commented:
re: 

> Have you evaluated if the application still needs this rule at all? I'm wondering if we can remove this code (& subsequent docs section.)

Currently the rules ensure that only layers with popups are selected. Given the state of the app that probably makes sense. I wouldn't want to remove it until we have more web maps to test with and can develop good behavior for e.g. identified raster layers.

> It appears this PR introduces support for identifying graphic overlays, I have a few questions here. Namely how do we support graphic overlays? More specifically,

The existing identify controller had code to support identifying graphics overlays. It never came up and was never really tested because the app doesn't currently support graphics overlays. The results of the graphics overlay query aren't consumed, so it would probably make sense to exclude that operation, but I left it alone since it isn't directly related to supporting multiple identify results.

> Can we consider a different static string other than `'Subtitle expression'`? If this expression is tied to the layer, perhaps we simply go with `'subtitle'`? This might be an easier standard for our customers to adopt.

I agree it is probably more usable with a simpler title.

> I'm wondering if we should consider sticking this static string in a Configuration file? This way it is easier to change alongside any other web map configs.

I'll add that feature

esreli commented:
> Currently the rules ensure that only layers with popups are selected. Given the state of the app that probably makes sense.

Because the identify task [method](https://developers.arcgis.com/net/latest/wpf/api-reference/html/M_Esri_ArcGISRuntime_UI_Controls_GeoView_IdentifyLayersAsync_3.htm) accepts a parameter `bool returnPopupsOnly`, wouldn't specifying `true` enforce that rule by default? Maybe i'm wrong but this seems to invalidate the need for this rule.

> I wouldn't want to remove it until we have more web maps to test with and can develop good behavior for e.g. identified raster layers.

Good point, though I will suggest that this is an issue we should tackle regardless of whether this rule is in place. It seems as though even with the rule in place, we don't know how the app would respond if an identified layer is a raster layer with pop-ups configured (TBH, I don't know if this is even possible).

> The existing identify controller had code to support identifying graphics overlays.

I think we should get rid of it, is that reasonable and are you okay with that?

nCastle1 committed:
Update doc, make more settings configurable

nCastle1 committed:
Remove graphics overlay identification code

This code wasn't being exercised

nCastle1 committed:
Remove identify app rule code

nCastle1 committed:
Improve UI behavior

Behavior is much better when the feature subtitle is not available, and feature graphics appearance is improved.

nCastle1 committed:
Fixes bug where scrollbar could reappear

nCastle1 committed:
Adds missed null check

nCastle1 commented:
@esreli The documentation around `returnPopupsOnly` is somewhat confusing. It doesn't filter for only results with popups, it filters the results to only include popups (and not the associated geoelement/ArcGISFeature).

nCastle1 committed:
Adds mention of new setting to docs

esreli commented:
> @esreli The documentation around `returnPopupsOnly` is somewhat confusing. It doesn't filter for only results with popups, it filters the results to only include popups (and not the associated geoelement/ArcGISFeature).

A pop-up cannot be created unless it has been enabled (and configured) on the service-side. This creates a pattern where the service determines what data can be identified rather than the client. Also, the underlying geoelement/ArcGISFeature is still accessible via the pop-up. Isn't this the behavior we are hoping to enforce?

esreli commented on RELEASE.md:
```diff
@@ -8,7 +8,8 @@
 * Updates the code called at sign out to revoke credentials from the server, when possible, and delete them locally.
-* Adds support for identifying mulitple items on the map simultaneously. A unique interface gives you enough context to choose a result, with context display configurable via the web map.
+* Adds support for identifying mulitple items on the map simultaneously and displaying them in a list.
+    * An arcade expression defined per layer is used to create subtitles that help you differentiate identified features.
```
I believe Arcade is a proper noun and should be capitalized. Also, this should be optional, no?

```suggestion
    * An Arcade expression defined per layer can be used to create subtitles that help you differentiate identified features.
```

esreli commented on docs/README.md:
```diff
@@ -366,33 +365,13 @@ The app allows a user to authenticate against a portal as well as use social cre
-The subtitle is computed via an Arcade expression defined in the pop-up configuration's *Attribute Expressions* collection. To be displayable in the app, the attribute expression must have the alias or title set to `Subtitle expression`.
+The subtitle is computed via an [Arcade expression](https://developers.arcgis.com/arcade/) defined in the pop-up configuration's *Attribute Expressions* collection. The app will only display a subtitle if the attribute expression has the alias or title set to match the **PopupExpressionForSubtitle** key defined in the app's settings.
 
 > **Note**, the `Subtitle expression` is found by an exact string match. Do not change the title when localizing the web map or for any other reason.
```
I believe this should be removed.
```suggestion
```

esreli commented on src/DataCollection.Shared/Properties/Configuration.xml:
```diff
@@ -6,6 +6,8 @@
   <DefaultZoomScale>5000</DefaultZoomScale>
   <AppClientID>h3em0ifYNGfz3uHX</AppClientID>
   <RedirectURL>data-collection://auth</RedirectURL>
+  <PopupExpressionForSubtitle>Subtitle expression</PopupExpressionForSubtitle>
```
can we change this to something a bit more simple? `'subtitle'`?

```suggestion
  <PopupExpressionForSubtitle>subtitle</PopupExpressionForSubtitle>
```

esreli reviewed (commented):
I was unable to properly review this PR due to an identify error discovered while using the WPF build.

### Identify Error

When I perform an identify function on the map view, i'm thrown an error. This is happening in online & offline maps.

![Screen Shot 2020-10-05 at 2 59 25 PM](https://user-images.githubusercontent.com/33433113/95136653-61b9a580-071b-11eb-8f08-fc7d9d15256f.png)

nCastle1 commented:
> 
> 
> > @esreli The documentation around `returnPopupsOnly` is somewhat confusing. It doesn't filter for only results with popups, it filters the results to only include popups (and not the associated geoelement/ArcGISFeature).
> 
> A pop-up cannot be created unless it has been enabled (and configured) on the service-side. This creates a pattern where the service determines what data can be identified rather than the client. Also, the underlying geoelement/ArcGISFeature is still accessible via the pop-up. Isn't this the behavior we are hoping to enforce?

I missed that the geoelement is still accessible through the pop-up. The existing code was checking the geoelement on the identify result directly. I'll update the code to set `returnPopupsOnly` true and get the feature from the popup.

nCastle1 committed:
Change how identify works

nCastle1 committed:
Doc and settings updates

nCastle1 committed:
Fixes UI issues for UWP

nCastle1 committed:
Code formatting

esreli commented on src/DataCollection.UWP/Views/Cards/IdentifyResultCard.xaml:
```diff
@@ -9,12 +9,20 @@
+            <ColumnDefinition Width="*" />
+        </Grid.ColumnDefinitions>
+        <Border Padding="8,0,0,0" Style="{StaticResource CardTopBarStyle}">
+            <BitmapIcon Style="{StaticResource PathIconWorkaroundStyle}" UriSource="Assets/Calcite/list-32.png" />
```
Much better!

esreli reviewed (commented):
This is looking great, I only have one comment.

### Bookmarks

It appears bookmarks have vanished. Do you know what's up here?

<img width="496" alt="Screen Shot 2020-10-06 at 9 59 10 AM" src="https://user-images.githubusercontent.com/33433113/95235539-96306e80-07ba-11eb-8f3d-aae811cb576f.png">

ZackAllen commented on src/DataCollection.Shared/Properties/Settings.cs:
```diff
@@ -265,5 +273,27 @@ private static Settings DeserializeSettings(object o)
+        {
+            if (_settingsInstance.MaxIdentifyResultsPerLayer < 1)
+            {
+                _settingsInstance.MaxIdentifyResultsPerLayer = 8;
```
Maybe you should add a comment to explain that this is setting a default when the setting is invalid. 8 seems random without context.

ZackAllen reviewed (approved):
LGTM

nCastle1 committed:
Adds comment
